unblob/handlers/compression/gzip.py

```diff
@@ -0,0 +1,235 @@
+import gzip
+import io
+import os
+import re
+from binascii import unhexlify
+from typing import List, Optional
+
+from structlog import get_logger
+
+from unblob.file_utils import find_first_backwards
+
+from ...models import StructHandler, ValidChunk
+
+logger = get_logger()
+
+
+# https://datatracker.ietf.org/doc/html/rfc1952
+class GZIPHandler(StructHandler):
+    NAME = "gzip"
+
+    YARA_RULE = r"""
+    strings:
+        // id1 & id2
+        // compression method (0x8 = DEFLATE)
+        // flags, 00011111 (0x1f) is the highest since the first 3 bits are reserved
+        // unix time
+        // eXtra FLags
+        // Operating System (RFC1952 describes 0-13, or 255)
+        $gzip_magic = /\x1f\x8b\x08[\x00-\x1f][\x00-\xff]{4}[\x02\x04][\x00-\x0c\xff]/
+    condition:
+        $gzip_magic
+    """
+
+    C_DEFINITIONS = r"""
+        struct gzip_struct {
+            char id1;
+            char id2;
+            char compression_method;
+            uint8 flags;
+            uint32 modification_time;
+            char extra_flags;
+            uint8 os; // Operating system
+        }
+
+        struct gzip_footer {
+            uint32 crc;
+            uint32 decompressed_length;
+        }
+    """
+
+    HEADER_STRUCT = "gzip_struct"
+
+    def calculate_chunk(
+        self, file: io.BufferedIOBase, start_offset: int
+    ) -> Optional[ValidChunk]:
+
+        try:
+            # If this is just a valid gzip file, we should be able to just read it into a
+            # gzip.py GzipFile, and just get the end offset like that. Easy!
+
+            with gzip.open(file) as g:
+                g.seek(0, os.SEEK_END)
```

Comment:
seek is using `self.read()` until the file is exhausted. If we would use `_GzipReader.read()` directly, we can stop at the proper offset, because `_read_gzip_header` will raise the `gzip.BadGzipFile` and we can stop there.

Calculating the correct offset can be copied from `_compression.DecompressReader.seek()` method, it's doing some adjustments and returning the correct offset.